crates/ty_python_semantic/resources/mdtest/typed_dict.md

```diff
@@ -127,6 +136,39 @@ dangerous(alice)
 reveal_type(alice["name"])  # revealed: Unknown
 ```
 
+## Key-based access
+
+```py
+from typing import TypedDict, Final, Literal, Any
+
+class Person(TypedDict):
+    name: str
+    age: int | None
+
+NAME_FINAL: Final = "name"
+AGE_FINAL: Final[Literal["age"]] = "age"
+
+def _(person: Person, literal_key: Literal["age"], union_of_keys: Literal["age", "name"], str_key: str, unknown_key: Any) -> None:
+    reveal_type(person["name"])  # revealed: str
+    reveal_type(person["age"])  # revealed: int | None
+
+    reveal_type(person[NAME_FINAL])  # revealed: str
+    reveal_type(person[AGE_FINAL])  # revealed: int | None
+
+    reveal_type(person[literal_key])  # revealed: int | None
+
+    reveal_type(person[union_of_keys])  # revealed: int | None | str
+
+    # error: [invalid-key] "Invalid key access on TypedDict `Person`: Unknown key "non_existing""
+    reveal_type(person["non_existing"])  # revealed: Unknown
+
+    # error: [invalid-key] "TypedDict `Person` can not be indexed with a key of type `str`"
+    reveal_type(person[str_key])  # revealed: Unknown
```

Comment:
We can see a few ecosystem hits due to this behavior. But it's relatively clear in the spec that this should be disallowed:

> A key that is not a literal should generally be rejected, since its value is unknown during type checking, and thus can cause some of the above violations.

> Type checkers should allow [final names](https://typing.python.org/en/latest/spec/qualifiers.html#uppercase-final) with string values to be used instead of string literals in operations on TypedDict objects.

> Similarly, an expression with a suitable [literal type](https://typing.python.org/en/latest/spec/literal.html#literal) can be used instead of a literal value